kore/src/Kore/Profiler/Data.hs

```diff
@@ -0,0 +1,114 @@
+{- |
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+-}
+module Kore.Profiler.Data
+    ( MonadProfiler (..)
+    , nullProfileDuration
+    , profileDurationEvent
+    , profileDurationStartEnd
+    , ProfileEvent (..)
+    ) where
+
+import           Control.Monad.Reader
+import qualified Control.Monad.State.Strict as Strict
+import           Control.Monad.Trans.Accum
+import           Control.Monad.Trans.Except
+import           Control.Monad.Trans.Identity
+import           Control.Monad.Trans.Maybe
+
+import           Debug.Trace
+                 ( traceEventIO )
+import           ListT
+                 ( ListT (..) )
+import qualified ListT
+                 ( gather, scatter )
+import           System.CPUTime
+                 ( getCPUTime )
+
+{- Monad that can also handle profiling events.
+-}
+class Monad profiler => MonadProfiler profiler where
```

Comment:
I think think this class in mainly needed because of `getCPUTime` below, right? Otherwise, we could use `traceEvent` anywhere.

Reply:
Mostly. The most important problem is that, although `traceEvent` should record a timestamp, it somehow fails to do that properly, so we're recording ours.

As a secondary problem, recording both the start and the end timestamp of an event in the same log statement is probably more reliable than recording two separate events.